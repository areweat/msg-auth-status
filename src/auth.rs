#[derive(Clone, Debug, Default, PartialEq)]
pub struct SmtpAuthResult<'hdr> {
    pub code: SmtpAuthResultCode,
    pub smtp_auth: Option<&'hdr str>,
    pub smtp_mailfrom: Option<&'hdr str>,
    pub raw: Option<&'hdr str>,
}

impl<'hdr> SmtpAuthResult<'hdr> {
    pub(crate) fn set_smtp(&mut self, prop: &ptypes::AuthSmtp<'hdr>) -> bool {
        match prop {
            ptypes::AuthSmtp::MailFrom(val) => self.smtp_mailfrom = Some(val),
            ptypes::AuthSmtp::Auth(val) => self.smtp_auth = Some(val),
            _ => {}
        }
        true
    }
}

/// SMTP AUTH Result Codes - s.2.7.4
/// This SMTP Authentication (not DKIM)
//#[derive(Debug, Default, EnumString, StrumDisplay)]
//#[strum(serialize_all = "lowercase", ascii_case_insensitive)]
#[derive(Clone, Debug, Default, PartialEq)]
pub enum SmtpAuthResultCode {
    #[default]
    Unknown,
    /// SMTP authentication was not attempted.
    NoneSmtp,
    /// The SMTP client authenticated to the server
    Pass,
    /// The SMTP client attempted to authenticate but was not successful
    Fail,
    /// The SMTP client attempted to authenticate but was not able to complete
    /// the attempt due to some error that is likely transient in nature
    TempError,
    /// The SMTP client attempted to authenticate but was not able to complete
    /// the attempt due to some error that is likely not transient in nature
    PermError,
}

pub mod ptypes;
pub use ptypes::AuthProperty;
