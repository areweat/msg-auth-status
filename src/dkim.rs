//! Public DKIM types exported by this library

#[derive(Clone, Debug, Default, PartialEq)]
pub struct DkimResult<'hdr> {
    pub code: DkimResultCode,
    pub reason: Option<&'hdr str>,
    pub header_d: Option<&'hdr str>,
    pub header_i: Option<&'hdr str>,
    pub header_b: Option<&'hdr str>,
    pub header_a: Option<DkimAlgorithm<'hdr>>,
    pub header_s: Option<&'hdr str>,
    pub raw: Option<&'hdr str>,
}

impl<'hdr> DkimResult<'hdr> {
    pub(crate) fn set_header(&mut self, prop: &ptypes::DkimHeader<'hdr>) -> bool {
        match prop {
            ptypes::DkimHeader::D(val) => self.header_d = Some(val),
            ptypes::DkimHeader::I(val) => self.header_i = Some(val),
            ptypes::DkimHeader::B(val) => self.header_b = Some(val),
            ptypes::DkimHeader::A(val) => self.header_a = Some(val.clone()),
            ptypes::DkimHeader::S(val) => self.header_s = Some(val),
            _ => {}
        }
        true
    }
}

/// DKIM Result Codes - s.2.7.1
//#[derive(Debug, Default, EnumString, StrumDisplay)]
//#[strum(serialize_all = "lowercase", ascii_case_insensitive)]
#[derive(Clone, Debug, Default, PartialEq)]
pub enum DkimResultCode {
    #[default]
    Unknown,
    /// The message was not signed.
    NoneDkim,
    /// The message was signed, the signature or signatures were
    /// acceptable to the ADMD, and the signature(s) passed verification
    /// tests.
    Pass,
    /// The message was signed and the signature or signatures were acceptable
    /// to the ADMD, but they failed the verification test(s).
    Fail,
    /// The message was signed, but some aspect of the signature or
    /// signatures was not acceptable to the ADMD.
    Policy,
    /// The message was signed, but the signature or signatures
    /// contained syntax errors or were not otherwise able to be
    /// processed.  This result is also used for other failures not
    /// covered elsewhere in this list.
    Neutral,
    /// The message could not be verified due to some error that
    /// is likely transient in nature, such as a temporary inability to
    /// retrieve a public key.  A later attempt may produce a final
    /// result.
    TempError,
    /// The message could not be verified due to some error that
    /// is unrecoverable, such as a required header field being absent.
    /// A later attempt is unlikely to produce a final result.
    PermError,
}

#[derive(Clone, Debug, Default, PartialEq)]
pub enum DKimCanonicalization<'hdr> {
    #[default]
    /// simple/simple & simple algorithm tolerates almost no modification
    Simple,
    /// relaxed/simple & relaxed algorithm tolerates common modifications such
    /// as whitespace replacement and header field line rewrapping.    
    Relaxed,
    /// Unknown RFC does not define
    Unknown(&'hdr str),
}

#[derive(Clone, Debug, Default, PartialEq)]
pub enum DkimTimestamp<'hdr> {
    #[default]
    Unknown,
    Raw(&'hdr str),
}

#[derive(Clone, Debug, PartialEq)]
#[allow(non_camel_case_types)]
pub enum DkimAlgorithm<'hdr> {
    /// Do not use
    Rsa_Sha1,
    /// Widely supported
    Rsa_Sha256,
    /// Please support this - not widely supported yet
    Ed25519_Sha256,
    Unknown(&'hdr str),
}
#[derive(Clone, Debug, PartialEq)]
pub enum DkimVersion<'hdr> {
    /// RFC just says this should be used
    One,
    /// Something else outside RFC
    Unknown(&'hdr str),
}

pub mod ptypes;
pub use ptypes::DkimProperty;
