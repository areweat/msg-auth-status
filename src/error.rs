//! Public errors

/// Errors relating to parsing Authentication-Results Header
#[derive(Clone, Debug, PartialEq)]
pub enum AuthResultsError {
    /// Unknown parsing error
    Parse,
    /// Host parsing error
    ParseHost(String),
    /// Bug
    ParsePtypeBugGating,
    /// Bug
    ParsePtypeBugInvalidProperty,
    /// Bug
    ParsePtypeBugPropertyGating,
    /// Invalid ptype Encountered
    ParsePtypeInvalidPtype,
    /// Invalid dkim method Result Code
    InvalidDkimResult(String),
    /// Invalid spf method Result Code
    InvalidSpfResult(String),
    /// Invalid iprev method Result Code
    InvalidIpRevResult(String),
    /// Was not a valid ptype/property per IANA and strict validation was used
    InvalidProperty,
    /// Invalid auth method Result code
    InvalidSmtpAuthResult(String),
    /// Invalid stage in result
    InvalidResultStage,
    /// Invalid version - Only 1 allowed
    InvalidVersion,
    /// No associated version found when required
    NoAssociatedVersion,
    /// No associated policy found when defined
    NoAssociatedPolicy,
    /// No assicited reason found when defined
    NoAssociatedReason,
    /// No hostname found that is required
    NoHostname,
    /// Err
    ParsePtypeNoMethodResult,
    /// Bug property keys not implemented for ptype
    PropertiesNotImplemented,
    /// Bug property values not implemented for ptype
    PropertyValuesNotImplemented,
    /// Run-away auth method property key
    RunAwayAuthPropertyKey,
    /// Run-away auth method property value
    RunAwayAuthPropertyValue,
    /// Run-away comment
    RunAwayComment,
    /// Run-away dkim method property key
    RunAwayDkimPropertyKey,
    /// Run-away dkim method property value
    RunAwayDkimPropertyValue,
    /// Run-away iprev method property key
    RunAwayIpRevPropertyKey,
    /// Run-away iprev method property value
    RunAwayIpRevPropertyValue,
    /// Run-away spf method property key
    RunAwaySpfPropertyKey,
    /// Run-away spf method property value
    RunAwaySpfPropertyValue,
    /// Unexpected forward slash
    UnexpectedForwardSlash,
    /// Bug
    ParseCurrentPushNotImplemented,
}
