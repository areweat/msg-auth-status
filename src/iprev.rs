//! Method iprev Result

/// Parsed iprev=..
#[derive(Clone, Debug, Default, PartialEq)]
pub struct IpRevResult<'hdr> {
    /// iprev resultcode
    pub code: IpRevResultCode,
    /// reason = ..
    pub reason: Option<&'hdr str>,
    /// iprev policy.iprev = ..
    pub policy_iprev: Option<&'hdr str>,
    /// unparsed
    pub raw: Option<&'hdr str>,
}

impl<'hdr> IpRevResult<'hdr> {
    pub(crate) fn set_policy(&mut self, prop: &ptypes::IpRevPolicy<'hdr>) -> bool {
        match prop {
            ptypes::IpRevPolicy::IpRev(val) => self.policy_iprev = Some(val),
            ptypes::IpRevPolicy::Unknown(_key, _val) => {} // ignore RFC unknowns
        }
        true
    }
    pub(crate) fn set_smtp(&mut self, _prop: &ptypes::IpRevSmtp<'hdr>) -> bool {
        // ignore these rfc breaking things
        false
    }
}

/// IpRev Result Codes - s.2.7.3
#[derive(Clone, Debug, Default, PartialEq)]
pub enum IpRevResultCode {
    /// Result code not seen
    #[default]
    Unknown,
    /// The DNS evaluation succeeded, i.e., the "reverse" and
    /// "forward" lookup results were returned and were in agreement.
    Pass,
    /// The DNS evaluation failed.  In particular, the "reverse" and
    /// "forward" lookups each produced results, but they were not in
    /// agreement, or the "forward" query completed but produced no
    /// result, e.g., a DNS RCODE of 3, commonly known as NXDOMAIN, or an
    /// RCODE of 0 (NOERROR) in a reply containing no answers, was
    /// returned.
    Fail,
    /// The DNS evaluation could not be completed due to some
    /// error that is likely transient in nature, such as a temporary DNS
    /// error, e.g., a DNS RCODE of 2, commonly known as SERVFAIL, or
    /// other error condition resulted.  A later attempt may produce a
    /// final result.
    TempError,
    /// The DNS evaluation could not be completed because no PTR
    /// data are published for the connecting IP address, e.g., a DNS
    /// RCODE of 3, commonly known as NXDOMAIN, or an RCODE of 0 (NOERROR)
    /// in a reply containing no answers, was returned.  This prevented
    /// completion of the evaluation.  A later attempt is unlikely to
    /// produce a final result.
    PermError,
}

pub mod ptypes;
pub use ptypes::IpRevProperty;
