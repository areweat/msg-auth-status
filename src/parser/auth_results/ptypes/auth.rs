//! auth ptype and it's properties
//!
//! See IANA Assignments
//! https://www.iana.org/assignments/email-auth/email-auth.xhtml
//!
//! And Auth in RFC 8601 s. 2.7

#[derive(Debug)]
pub enum AuthProperty<'hdr> {
    SmtpAuth(&'hdr str),
    MailFrom(&'hdr str),
}

#[derive(Debug)]
pub enum AuthPtype {
    Smtp,
    SmtpDotAuth,
    Mail,
    MailDotAuth,
}
