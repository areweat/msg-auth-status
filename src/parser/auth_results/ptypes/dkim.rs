//! Parsing dkim property types & values

use super::*;

pub mod dkim_property_key;
pub mod dkim_property_value;
