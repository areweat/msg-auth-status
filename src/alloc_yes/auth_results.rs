//! Allocating DkimResultsHandler, SpfResultsHandler, AuthResultsHandler, IpRevResultsHandler

use crate::auth::SmtpAuthResult;
use crate::dkim::DkimResult;
use crate::iprev::IpRevResult;
use crate::spf::SpfResult;

use crate::auth_results::*;

use crate::error::AuthResultsError;

/// Parsed Authentication-Results
#[derive(Clone, Debug, Default, PartialEq)]
pub struct AuthenticationResults<'hdr> {
    /// Relevant host to this record
    pub host: Option<HostVersion<'hdr>>,
    /// Parsed auth = .. records
    pub smtp_auth_result: Vec<SmtpAuthResult<'hdr>>,
    /// Parsed spf = .. records
    pub spf_result: Vec<SpfResult<'hdr>>,
    /// Parsed dkim = .. records
    pub dkim_result: Vec<DkimResult<'hdr>>,
    /// Parsed iprev = .. records
    pub iprev_result: Vec<IpRevResult<'hdr>>,
    /// Whether none was encountered denoting no result
    pub none_done: bool,
    /// Unparsed raw
    pub raw: Option<&'hdr str>,
    /// Parsing errors if any
    pub errors: Vec<AuthResultsError<'hdr>>,
}

/// Allocating type for parsed all Authentication-Results in email
#[cfg(any(feature = "alloc", feature = "std"))]
#[derive(Debug, Default)]
pub struct MessageAuthStatus<'hdr> {
    /// Authentication-Results
    auth_results: Vec<AuthenticationResults<'hdr>>,
}

/// TODO: Fix errors
#[derive(Debug, PartialEq)]
pub enum MessageAuthStatusError {}

impl<'hdr> MessageAuthStatus<'hdr> {
    /// Parse all Authentication-Results into allocating Vec from external mail_parser::Message
    #[cfg(feature = "mail_parser")]
    pub fn from_mail_parser(
        msg: &'hdr mail_parser::Message<'hdr>,
    ) -> Result<Self, MessageAuthStatusError> {
        let mut new_self = Self {
            auth_results: vec![],
        };

        new_self.auth_results = msg
            .header_values("Authentication-Results")
            .map(|mh| mh.into())
            .collect();

        Ok(new_self)
    }
}

#[cfg(test)]
#[cfg(feature = "mail_parser")]
mod test {
    use super::*;
    use insta::assert_debug_snapshot;
    use rstest::rstest;
    use std::{fs::File, io::Read, path::PathBuf};

    fn load_test_data(file_location: &str) -> Vec<u8> {
        let mut file = File::open(file_location).unwrap();
        let mut data: Vec<u8> = vec![];
        file.read_to_end(&mut data).unwrap();
        data
    }

    #[rstest]
    fn from_mail_parser(#[files("test_data/*.eml")] file_path: PathBuf) {
        let new_snapshot_path = file_path.with_extension("snap");

        insta::with_settings!({snapshot_path => new_snapshot_path}, {
            insta::allow_duplicates! {
                let raw  = load_test_data(file_path.to_str().unwrap());
                let parser = mail_parser::MessageParser::default();
                let parsed_message = parser.parse(&raw).unwrap();
                let status = MessageAuthStatus::from_mail_parser(&parsed_message).unwrap();
                assert_debug_snapshot!(&status);
            }
        });
    }
}
