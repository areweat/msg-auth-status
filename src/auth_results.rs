//! Parsed Authetication-Results

use crate::auth::AuthProperty;
use crate::dkim::DkimProperty;
use crate::iprev::IpRevProperty;
use crate::spf::SpfProperty;

/// Host and version relating to the results
#[derive(Clone, Debug, PartialEq)]
pub struct HostVersion<'hdr> {
    /// Host
    pub host: &'hdr str,
    /// Version
    pub version: Option<u32>,
}

/// Auth-Result properties
#[derive(Clone, Debug, PartialEq)]
pub enum Prop<'hdr> {
    /// auth method properties
    Auth(AuthProperty<'hdr>),
    /// dkim method properties
    Dkim(DkimProperty<'hdr>),
    //dmarc method properties
    //Dmarc(DmarcProperty<'hdr>),
    /// iprev method properties
    IpRev(IpRevProperty<'hdr>),
    /// spf method properties
    Spf(SpfProperty<'hdr>),
    /// unknown method properties
    Unknown(UnknownProperty<'hdr>),
}

/// Unknown method properties
#[derive(Clone, Debug, PartialEq)]
pub struct UnknownProperty<'hdr> {
    /// Unknown ptype
    ptype: &'hdr str,
    /// Unknown property key
    pkey: &'hdr str,
    /// Unknown property value
    pval: &'hdr str,
}
