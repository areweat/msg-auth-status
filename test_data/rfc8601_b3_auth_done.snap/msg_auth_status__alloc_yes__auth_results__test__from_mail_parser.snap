---
source: src/alloc_yes/auth_results.rs
expression: "&status"
---
MessageAuthStatus {
    auth_results: [
        AuthenticationResults {
            host: Some(
                HostVersion {
                    host: "example.com",
                    version: None,
                },
            ),
            smtp_auth_result: [],
            spf_result: [
                SpfResult {
                    code: Pass,
                    reason: None,
                    smtp_mailfrom: Some(
                        "example.net",
                    ),
                    smtp_helo: None,
                    raw: Some(
                        "spf=pass smtp.mailfrom=example.net",
                    ),
                },
            ],
            dkim_result: [],
            iprev_result: [],
            none_done: false,
            raw: Some(
                "example.com;\n          spf=pass smtp.mailfrom=example.net",
            ),
            error: None,
        },
    ],
}
